lib/puppet/provider/package/puppetserver_gem.rb

```diff
@@ -13,53 +16,111 @@
 
   has_feature :versionable, :install_options, :uninstall_options
 
-  commands :puppetservercmd => "/opt/puppetlabs/bin/puppetserver"
-
-  # the HOME variable is lost to the puppetserver script and needs to be
-  # injected directly into the call to `execute()`
-  CMD_ENV = {:custom_environment => {:HOME => ENV['HOME']}}
+  confine :feature => :hocon
+  commands :puppetservercmd => '/opt/puppetlabs/bin/puppetserver'
 
+  # The HOME variable is lost to the puppetserver script
+  #  and needs to be injected directly into the call to `execute()`
+  # When doing so, restore :failonfail and :combine to their defaults
+  #  as per the documentation in lib/puppet/util/execution.rb
+  EXEC_OPTS = { :failonfail => true, :combine => true, :custom_environment => { :HOME => ENV['HOME'] } }
 
   def self.gemlist(options)
-    gem_list_command = [command(:puppetservercmd), "gem", "list"]
+    gem_list_command = [command(:puppetservercmd), 'gem', 'list']
 
     if options[:local]
-      gem_list_command << "--local"
+      gem_list_command << '--local'
     else
-      gem_list_command << "--remote"
+      gem_list_command << '--remote'
     end
+
     if options[:source]
-      gem_list_command << "--source" << options[:source]
+      gem_list_command << "--source #{options[:source]}"
     end
+
     if name = options[:justme]
-      gem_list_command << "^" + name + "$"
+      gem_regex = '\A' + name + '\z'
+      gem_list_command << gem_regex
     end
 
-    begin
-      list = execute(gem_list_command, CMD_ENV).lines.
-          map {|set| gemsplit(set) }.
-          reject {|x| x.nil? }
-    rescue Puppet::ExecutionFailure => detail
-      raise Puppet::Error, "Could not list gems: #{detail}", detail.backtrace
+    if options[:local]
+      list = execute_rubygems_list_command(gem_regex)
+    else
+      begin
+        list = execute(gem_list_command, EXEC_OPTS)
+      rescue Puppet::ExecutionFailure => detail
+        raise Puppet::Error, _("Could not list gems: %{detail}") % { detail: detail }, detail.backtrace
+      end
     end
 
+    # When `/tmp` is mounted `noexec`, `puppetserver gem list` will output *** LOCAL GEMS ***
+    #  causing gemsplit to output: Warning: Could not match *** LOCAL GEMS ***
+    gem_list = list
+               .lines
+               .select { |x| x =~ /^(\S+)\s+\((.+)\)/ }
+               .map { |set| gemsplit(set) }
+
     if options[:justme]
-      return list.shift
+      return gem_list.shift
     else
-      return list
+      return gem_list
     end
   end
 
+  # The puppetserver gem cli commands are particularly slow as they start a JVM.
+  # Instead, for the often-executed list command, use the rubygems library from the puppet ruby,
+  #  setting the GEM_HOME and GEM_PATH to the values in the puppetserver configuration file.
+  # The rubygems library does not have access to java platform gems, for example: json (1.8.3 java),
+  #  but java platform gems should not be managed, by design.
```

Comment:
Does this mean that it won't list JRuby versions of gems that get installed (i.e. gems with native Java extensions)? Is that acceptable?

Reply:
(Thanks, I pinged Modules ...)

It means java platform gems will not be manageable by this provider ... although they are manageable via the underlying `puppetserver gem` command. For example, given:

```
[root@pe-201815-master ~]# puppetserver gem list

*** LOCAL GEMS ***

deep_merge (1.0.1)
did_you_mean (default: 1.0.1)
facter (2.5.1)
fast_gettext (1.1.2, 1.1.1)
gettext (3.2.2)
hiera-eyaml (2.1.0)
highline (1.6.21)
hocon (1.2.5, 1.1.3)
jar-dependencies (default: 0.3.12)
jruby-openssl (0.9.21 java)
jruby-readline (1.2.2 java)
json (1.8.3 java)
locale (2.1.2)
minitest (default: 5.4.1)
multi_json (1.13.1)
net-telnet (default: 0.1.1)
power_assert (default: 0.2.3)
psych (2.2.4 java)
puppetserver-ca (1.1.2)
rake (default: 10.4.2)
rdoc (default: 4.2.0)
semantic_puppet (0.1.3, 0.1.2)
test-unit (default: 3.1.1)
text (1.3.1)
trollop (2.9.9, 2.1.2)
```

Only the following could not be upgraded or uninstalled by this provider:

```
jruby-openssl (0.9.21 java)
jruby-readline (1.2.2 java)
json (1.8.3 java)
psych (2.2.4 java)
```

That seems acceptable and reasonable.